example/Binary.hs

```diff
@@ -0,0 +1,389 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Binary (
+  Bit (..),
+  LogicBit (..),
+  _LogicO,
+  _LogicI,
+  _LogicO',
+  _LogicI',
+  Binary,
+  zeroo,
+  poso,
+  gtlo,
+  binaryo,
+  addo,
+  subo,
+  lesso,
+  mulo,
+  divo,
+  example,
+) where
+
+import Control.Lens (Prism, from)
+import Control.Lens.TH (makePrisms)
+import Control.Monad (void)
+import Data.Bifunctor (bimap)
+import Data.Function ((&))
+import Data.Tagged (Tagged)
+import GHC.Generics (Generic)
+
+import Core
+import Goal
+import LogicalBase
+import Match
+import TH
+
+data Bit = O | I deriving (Eq, Show, Generic)
+
+makeLogic ''Bit
+deriving instance Show LogicBit
+
+makePrisms ''LogicBit
+
+_LogicO'
+  :: Prism
+      (Tagged (o, i) LogicBit)
+      (Tagged (o', i) LogicBit)
+      (Tagged o ())
+      (Tagged o' ())
+_LogicO' = from _Tagged . _LogicO . _Tagged
+
+_LogicI'
+  :: Prism
+      (Tagged (o, i) LogicBit)
+      (Tagged (o, i') LogicBit)
+      (Tagged i ())
+      (Tagged i' ())
+_LogicI' = from _Tagged . _LogicI . _Tagged
+
+type Binary = [Bit]
+
+instance Num Binary where
+  fromInteger 0 = []
+  fromInteger 1 = [I]
+  fromInteger x | x > 1 = bit : fromInteger (x `div` 2)
+   where
+    bit
+      | even x = O
+      | otherwise = I
+  fromInteger _ = error "Binary numbers cannot be negative"
+
+  (+) = undefined
+  (*) = undefined
+  abs = id
+  signum = undefined
+  negate = undefined
+
+zeroo :: Term Binary -> Goal ()
+zeroo n = n === Value LogicNil
+
+cons :: (Logical a) => Term [a] -> Goal (Term a, Term [a])
+cons list = do
+  (x, xs) <- fresh
+  list === Value (LogicCons x xs)
+  return (x, xs)
+
+poso :: Term Binary -> Goal ()
+poso = void . cons
+
+gtlo :: Term Binary -> Goal ()
```

Comment:
What is "gtl"? Please, provide description in the Haddock documentation. Also would be great to reference corresponding section/definition in the Kiselyov's paper.

Reply:
this name comes from [the original implementation](https://okmij.org/ftp/Prolog/Arithm/pure-bin-arithm.prl). I guess it's supposed to mean “greater than 1”, where they use `l` to denote the 1 bit. this was confusing to me as well, but I decided to stick to the original implementation, at least initially.

I will add references to the paper once I finish reimplementing `logo`.